src/common/ui/code_extras/mode/javascript-mixed/javascript-mixed.js

```diff
@@ -0,0 +1,304 @@
+      // to ensure the ending backtick is parsed correctly.
+      forceJsModeToQuasi(stream, state.jsState);
+    }
+
+    function exitLocalModeAndTokenEndingBacktick(stream, state) {
+      dbg('Exiting local html/css mode...');
+      // parse the ending JS string template backtick in js mode
+      return jsMode.token(stream, state.jsState);
+    }
+
+    function tokenInLocalMode(modeToUse, stream, state) {
+      // parse the text in local mode
+      state.localMode = state.localMode || modeToUse;
+      state.localState = state.localState || CodeMirror.startState(state.localMode);
+      const style = state.localMode.token(stream, state.localState);
+      dbg(`  local ${modeToUse.name} mode token - `, stream.current());
+      return style;
+    }
+
+    function matchRule(rules, stream, state) {
+      for (const r of rules) {
+        if (r.curContext === (state.maybeLocalContext || '<start>')) {
+          // dbg('  rule:', r.curContext, r.matchFn.toString());
+          const matched = r.run(stream, state);
+          if (matched) { break; }
+        }
+      }
+    }
+
+    class Rule {
+      constructor(curContext, matchFn, nextContext, caseMatchedFn, caseNotMatchedFn) {
+        this.curContext = curContext;
+        this.matchFn = matchFn;
+        this.nextContext = nextContext;
+        this.caseMatchedFn = caseMatchedFn; // optional
+        this.caseNotMatchedFn = caseNotMatchedFn; // optional
+      }
+
+      run(stream, state) {
+        if (this.matchFn()) {
+          state.maybeLocalContext = this.nextContext;
+          if (state.maybeLocalContext == null) {
+            // local mode done, reset
+            state.localMode = null;
+            state.localState = null;
+          }
+          if (this.caseMatchedFn) { this.caseMatchedFn(); }
+          return true;
+        } // case rule transition criteria not matched
+        if (this.caseNotMatchedFn) {
+          this.caseNotMatchedFn();
+        } else { // default not matched logic: reset local mode matching
+          state.maybeLocalContext = null;
+        }
+        return false;
+      }
+    }
+
+    const cssMatcher = (function () {
+      const cssMode = CodeMirror.getMode(config, { name: 'css' });
+
+      function maybeToken(stream, state, jsTokStyle) {
+        let tokStyle = STYLE_PASS;
+        const tokTyp = state.jsState.lastType;
+        const tokStr = stream.current();
+
+        // define the transition rules to enter local CSS mode;
+        const rules = [
+          // <current-context>, <match-criteria>, <next-context-if-matched>,
+          // <optional-side-effects-if-matched>,
+          // <optional-side-effects-not-matched>
+          //
+          // - side-effects-not-matched: if not specified, defaulted to reset local mode matching
+
+          // for pattern GM_addStyle(`css-string`);
+          new Rule('<start>', () => tokTyp === 'variable' && tokStr === 'GM_addStyle',
+            'css-1'),
```

Comment:
This wrapping looks weird. Since there are many parameters it might make sense to use named props so the call becomes:

```js
new Rule({
  curContext: '<start>',
  nextContext: 'css-1',
  matchFn: () => tokTyp === 'variable' && tokStr === 'GM_addStyle',
}),
```

Reply:
Using named props would be clearer but the rule set would become more verbose, harder to glance.

```javascript
          new Rule({
            curContext: 'css-1',
            matchFn: () => tokTyp === '(' && tokStr === '(',
            nextContext: 'css-2'
          }),
          new Rule({
            curContext: 'css-2',
            matchFn: () => tokTyp === 'quasi', // if it's a string template
            nextContext: 'css-in',
            caseMatchedFn: () => prepareReparseStringTemplateInLocalMode(cssMode, stream, state)
          }),
          new Rule({
            curContext: 'css-in',
            matchFn: () => stream.peek() === '`', // if it hits ending backtick for string template
            nextContext: null, // then exiting local css mode
            caseMatchedFn: () => { tokStyle = exitLocalModeAndTokenEndingBacktick(stream, state); },
            caseNotMatchedFn: () => { tokStyle = tokenInLocalMode(stream, state) // else stay in local mode
          }),
```

Should I go ahead and make the changes?

Reply:
Well, to you it may look overly verbose but to me (or other people unacquainted with this code) the named props variant looks more meaningful. It's also easier to maintain/extend. Performance won't suffer because the object will be created just once.

Reply:
BTW you can drop the `fn` suffix to simplify the names as it's unnecessary.

Reply:
done